fiftyone/utils/csv.py

```diff
@@ -6,13 +6,230 @@
+        self.fields = fields
+        self.skip_missing_media = skip_missing_media
+        self.include_all_data = include_all_data
+
+        self._media_paths_map = None
+        self._rows_map = None
+        self._fields = None
+        self._filepaths = None
+        self._iter_filepaths = None
+        self._num_samples = None
+
+    def __iter__(self):
+        self._iter_filepaths = iter(self._filepaths)
+        return self
+
+    def __len__(self):
+        return self._num_samples
+
+    def __next__(self):
+        filepath = next(self._iter_filepaths)
+
+        row = self._rows_map.get(filepath, None)
+        if row is None:
+            return fos.Sample(filepath=filepath)
+
+        kwargs = {}
+        if isinstance(self._fields, dict):
+            for key, parser in self._fields.items():
+                value = row.get(key, "")
+                if value:
+                    kwargs[key] = parser(value)
+        elif self._fields is not None:
+            for key in self._fields:
+                value = row.get(key, "")
+                if value:
+                    kwargs[key] = value
+        else:
+            for key, value in row.items():
+                if value:
+                    kwargs[key] = value
+
+        return fos.Sample(filepath=filepath, **kwargs)
+
+    @property
+    def has_sample_field_schema(self):
+        return False
+
+    @property
+    def has_dataset_info(self):
+        return False
+
+    def setup(self):
+        media_paths_map = self._load_data_map(self.data_path, recursive=True)
+
+        rows_map = {}
+        fields = _parse_import_fields(self.fields, self.media_field)
+
+        media_field = self.media_field
+        if self.labels_path is not None and os.path.isfile(self.labels_path):
+            with open(self.labels_path, "r") as f:
+                reader = csv.DictReader(f)
+                if media_field not in reader.fieldnames:
+                    raise ValueError(
+                        "Media field '%s' not found in '%s'"
+                        % (media_field, self.labels_path)
+                    )
+
+                for row in reader:
+                    filename = row.pop(media_field, None)
+                    if filename is None or os.path.isabs(filename):
+                        filepath = filename
+                    else:
+                        filepath = media_paths_map.get(filename, None)
+
+                    if filepath is None:
+                        if self.skip_missing_media:
+                            continue
+                        else:
+                            raise ValueError(
+                                "Found row with no '%s' value" % media_field
+                            )
+
+                    rows_map[filepath] = row
+
+        filepaths = set(rows_map.keys())
+
+        if self.include_all_data:
+            filepaths.update(media_paths_map.values())
+
+        filepaths = self._preprocess_list(sorted(filepaths))
```

Comment:
By convention, `(max_samples=n, shuffle=True)` samples from entire dataset. This implementation achieves that.